stackslib/src/net/api/get_tenures_fork_info.rs

```diff
@@ -0,0 +1,361 @@
+    }
+}
+
+trait HexDeser: Sized {
+    fn try_from(hex: &str) -> Result<Self, HexError>;
+}
+
+macro_rules! impl_hex_deser {
+    ($thing:ident) => {
+        impl HexDeser for $thing {
+            fn try_from(hex: &str) -> Result<Self, HexError> {
+                $thing::from_hex(hex)
+            }
+        }
+    };
+}
+
+impl_hex_deser!(BurnchainHeaderHash);
+impl_hex_deser!(StacksBlockId);
+impl_hex_deser!(SortitionId);
+impl_hex_deser!(ConsensusHash);
+impl_hex_deser!(BlockHeaderHash);
+impl_hex_deser!(Hash160);
+
+#[derive(Clone, Default)]
+pub struct GetTenuresForkInfo {
+    pub start_sortition: Option<ConsensusHash>,
+    pub stop_sortition: Option<ConsensusHash>,
+}
+
+/// Decode the HTTP request
+impl HttpRequest for GetTenuresForkInfo {
+    fn verb(&self) -> &'static str {
+        "GET"
+    }
+
+    fn path_regex(&self) -> Regex {
+        Regex::new(&format!(
+            r#"^{RPC_TENURE_FORKING_INFO_PATH}/(?P<start>[0-9a-f]{{40}})/(?P<stop>[0-9a-f]{{40}})$"#
+        ))
+        .unwrap()
+    }
+
+    /// Try to decode this request.
+    /// There's nothing to load here, so just make sure the request is well-formed.
+    fn try_parse_request(
+        &mut self,
+        preamble: &HttpRequestPreamble,
+        captures: &Captures,
+        query: Option<&str>,
+        _body: &[u8],
+    ) -> Result<HttpRequestContents, Error> {
+        if preamble.get_content_length() != 0 {
+            return Err(Error::DecodeError(
+                "Invalid Http request: expected 0-length body".to_string(),
+            ));
+        }
+
+        let req_contents = HttpRequestContents::new().query_string(query);
+
+        let start_str = captures
+            .name("start")
+            .ok_or_else(|| {
+                Error::DecodeError("Failed to match path to start_sortition group".to_string())
+            })?
+            .as_str();
+        let stop_str = captures
+            .name("stop")
+            .ok_or_else(|| {
+                Error::DecodeError("Failed to match path to stop_sortition group".to_string())
+            })?
+            .as_str();
+        let start_sortition = ConsensusHash::from_hex(start_str).map_err(|_| {
+            Error::DecodeError("Invalid path: unparseable consensus hash".to_string())
+        })?;
+        let stop_sortition = ConsensusHash::from_hex(stop_str).map_err(|_| {
+            Error::DecodeError("Invalid path: unparseable consensus hash".to_string())
+        })?;
+        self.start_sortition = Some(start_sortition);
+        self.stop_sortition = Some(stop_sortition);
+
+        Ok(req_contents)
+    }
+
+    fn metrics_identifier(&self) -> &str {
+        RPC_TENURE_FORKING_INFO_PATH
+    }
+}
+
+impl TenureForkingInfo {
+    fn from_snapshot(
+        sn: &BlockSnapshot,
+        sortdb: &SortitionDB,
+        chainstate: &StacksChainState,
+    ) -> Result<Self, ChainError> {
+        let first_block_mined = if !sn.sortition {
+            None
+        } else {
```

Comment:
A couple of optimizations you can make:

* You can load up the sortitions for the start and end sortitions eagerly, and if they don't exist, then this can return 404. They'd still be returned on a 200 so you don't waste a sortition load.
* You can check to see if the start sortition is an ancestor of the end sortition, and return 409 (or something) to indicate that the arguments are inconsistent.

Reply:
I’d rather avoid too many optimizations in this PR, and just focus on getting correctness first. Optimizations can make easy follow on work.